enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+
+**Note**: the label `selector` below is defined by the [forwarder label selectors proposal][clf-labels]
+
+```yaml
+inputs:
+- name: InputPatchy
+  applications:
+    selector:
+      matchLabels: { app: patchy }
+  structured: json
+  schema: patchy
+- name: InputBulldog
+  applications:
+    selector:
+      matchLabels: { app: bulldog }
+  structured: json
+  schema: bulldog
+pipelines:
+- inputRefs: [ InputPatchy, InputBulldog ]
+  outputRefs: default
+```
+
+#### I want to forward structured and unstructured records to the same output
+
+Structured records will have the `structured` field, non-structured will have the `message` field.
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputMaybeJSON
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputMaybeJSON ]
+```
+
+#### I want a replacement for the defunct MERGE_JSON_LOG feature
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputMerged
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputMerged ]
+  modify:
+    merge: true
+```
+
+#### I want to  forward structured and unstructured logs from the same input to different outputs
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputJSON
+  type: fluentdForward
+  ...
+outputs:
+- name: OutputNotJSON
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputJSON ]
+  filter:
+    require: [structured]
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputNotJSON ]
+  filter:
+    reject: [structured]
+```
+
+#### I want to forward structured records to one output, and all records in unstructured form to another
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputJSON
+  type: fluentdForward
+  ...
+outputs:
+- name: OutputAllText
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputJSON ]
+  filter:
+    require: [structured]
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputAllText ]
+  modify:
+    remove: [structured] # Replaced with an unstructured `message` field.
+```
+
+### Implementation Details
+
+#### Elasticsearch version-specific workarounds.
+
+The existing Elasticsearch logging store requires "flattening" of the `kubernetes.labels` map:
+
+* Transform key:value object into array of "NAME=VALUE" strings.
+* Replace '.' with '_' in label names.
+
+This makes label keys difficult to use for a consumer.
+There is no automatic way to reverse the process since '.' and '_' are both legal characters in label names.
+
+The new forwarder will check the version of the Elasticsearch Operator (via labels or annotations, TBD).
+Flattening labels will be enabled if the ELO is identified as a 'legacy' version.
```

Comment:
I think this is going to be more dependent on the version of elasticsearch that is running, not the operator (unless the operator is planned to configure the pipelines)

Reply:
Yes, I'm not sure exactly how to do that. I'm hoping that either the CLO or ELO operator can check  whether the pipeline is installed or not, or that whatever does install the pipeline leaves a marker (annotation say) somewhere. Need to think that part through.